File: code/packages/rust/json-rpc/src/writer.rs
//! `MessageWriter` — writes Content-Length-framed JSON-RPC messages to a
//! `Write` byte stream.
//!
//! ## Writing Algorithm
//!
//! For each message:
//! 1. Serialise the `Message` to a JSON byte vector (`serde_json::to_vec`).
//! 2. Compute `payload.len()` — the byte length (NOT the character count).
//!    UTF-8 characters can be 1–4 bytes; `len()` gives the correct value.
//! 3. Write `"Content-Length: {n}\r\n\r\n"` followed by the payload bytes.
//! 4. Flush the writer so the bytes reach the OS buffer immediately.
//!
//! ## Why Flush?
//!
//! If the writer is a `BufWriter<Stdout>`, bytes are accumulated in a
//! user-space buffer until the buffer fills or the writer is explicitly
//! flushed. Without flushing after each message, the client may wait
//! indefinitely for the response. We flush after every write.
//!
//! ## Usage
//!
//! ```rust,no_run
//! use coding_adventures_json_rpc::{MessageWriter, message::{Message, Response}};
//! use std::io::{BufWriter, stdout};
//!
//! let mut writer = MessageWriter::new(BufWriter::new(stdout()));
//! let response = Response {
//!     id: serde_json::json!(1),
//!     result: Some(serde_json::json!({"ok": true})),
//!     error: None,
//! };
//! writer.write_message(&Message::Response(response)).unwrap();
//! ```

use crate::message::{message_to_value, Message};
use std::io::{self, Write};

/// Writes Content-Length-framed JSON-RPC messages to a `Write` destination.
pub struct MessageWriter<W: Write> {
    writer: W,
}

impl<W: Write> MessageWriter<W> {
    // -----------------------------------------------------------------------
    // Constructor
    // -----------------------------------------------------------------------

    /// Create a new `MessageWriter` wrapping the given writer.
    ///
    /// ```rust
    /// use coding_adventures_json_rpc::MessageWriter;
    /// use std::io::Cursor;
    ///
    /// let writer = MessageWriter::new(Cursor::new(Vec::new()));
    /// ```
    pub fn new(writer: W) -> Self {
        Self { writer }
    }

    // -----------------------------------------------------------------------
    // write_message — serialise and write a typed message
    // -----------------------------------------------------------------------

    /// Serialise a `Message` and write it with Content-Length framing.
    ///
    /// Flushes the underlying writer after writing so the bytes reach the
    /// client immediately.
    ///
    /// # Errors
    ///
    /// Returns `Err(io::Error)` if serialisation or writing fails.
    ///
    /// # Example
    ///
    /// ```rust
    /// use coding_adventures_json_rpc::{MessageWriter, message::{Message, Notification}};
    /// use std::io::Cursor;
    ///
    /// let mut out = Cursor::new(Vec::new());
    /// let mut writer = MessageWriter::new(&mut out);
    /// let notif = Notification { method: "initialized".to_string(), params: None };
    /// writer.write_message(&Message::Notification(notif)).unwrap();
    ///
    /// let bytes = out.into_inner();
    /// let text = std::str::from_utf8(&bytes).unwrap();
    /// assert!(text.starts_with("Content-Length: "));
    /// ```
    pub fn write_message(&mut self, message: &Message) -> io::Result<()> {
        // Convert the typed message to a serde_json Value, then serialise it.
        let value = message_to_value(message);
        let payload = serde_json::to_vec(&value)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
        self.write_raw(&payload)
    }

    // -----------------------------------------------------------------------
    // write_raw — write a raw byte payload with framing
    // -----------------------------------------------------------------------

    /// Write a raw byte payload with Content-Length framing.
    ///
    /// Useful when the caller has already serialised the payload or wants to
    /// write a custom JSON body.
    ///
    /// ```rust
    /// use coding_adventures_json_rpc::MessageWriter;
    /// use std::io::Cursor;
    ///
    /// let mut out = Cursor::new(Vec::new());
    /// let mut writer = MessageWriter::new(&mut out);
    /// let json = br#"{"jsonrpc":"2.0","method":"ping"}"#;
    /// writer.write_raw(json).unwrap();
    ///
    /// let bytes = out.into_inner();
    /// assert!(bytes.starts_with(b"Content-Length: "));
    /// ```
    pub fn write_raw(&mut self, payload: &[u8]) -> io::Result<()> {
        // byte_size, not character count — crucial for UTF-8 payloads with
        // multi-byte characters (e.g. file paths with Unicode).
        let n = payload.len();

        // Write the header: "Content-Length: N\r\n\r\n"
        write!(self.writer, "Content-Length: {}\r\n\r\n", n)?;

        // Write the payload bytes.
        self.writer.write_all(payload)?;

        // Flush so the bytes are immediately visible to the reader.
        self.writer.flush()
    }

    // -----------------------------------------------------------------------
    // into_inner — extract the underlying writer
    // -----------------------------------------------------------------------

    /// Consume the `MessageWriter` and return the underlying writer.
    ///
    /// Useful in tests to inspect the written bytes.
    pub fn into_inner(self) -> W {
        self.writer
    }
}

File: code/packages/elixir/json_rpc/lib/json_rpc/writer.ex
defmodule CodingAdventures.JsonRpc.Writer do
  @moduledoc """
  MessageWriter — writes Content-Length-framed JSON-RPC messages to a device.

  ## The Framing Protocol

  Every message is prefixed with an HTTP-style header block and a blank line:

  ```
  Content-Length: 47\\r\\n
  \\r\\n
  {"jsonrpc":"2.0","id":1,"result":{"ok":true}}
  ```

  Steps:
  1. Encode the message struct to a JSON binary (UTF-8).
  2. Compute `byte_size(json)` — this is the Content-Length value. We use
     `byte_size` not `String.length` because UTF-8 characters can be 1–4 bytes.
  3. Write: `"Content-Length: N\\r\\n\\r\\n"` followed by the JSON bytes.

  ## Usage

      writer = Writer.new(:stdio)
      :ok = Writer.write_message(writer, %Response{id: 1, result: %{"ok" => true}})

  ## In-Memory Testing

      {:ok, pid} = StringIO.open("")
      writer = Writer.new(pid)
      :ok = Writer.write_message(writer, %Notification{method: "initialized"})
      {_in, out} = StringIO.contents(pid)
      # out contains the framed message bytes
  """

  alias CodingAdventures.JsonRpc.{JsonCodec, Message}

  defstruct [:device]

  @type t :: %__MODULE__{device: any()}

  # ---------------------------------------------------------------------------
  # new/1
  # ---------------------------------------------------------------------------

  @doc """
  Create a new MessageWriter from an I/O device.

      writer = Writer.new(:stdio)
  """
  @spec new(any()) :: t()
  def new(device) do
    %__MODULE__{device: device}
  end

  # ---------------------------------------------------------------------------
  # write_message/2 — encode a typed message and write it with framing
  # ---------------------------------------------------------------------------

  @doc """
  Write a typed message struct to the device with Content-Length framing.

  Converts the message to a plain map, encodes it as JSON, computes the byte
  length, and writes the header + payload.

  Returns `:ok` on success or `{:error, reason}` on failure.

  ## Example

      :ok = Writer.write_message(writer, %Response{id: 1, result: nil})
  """
  @spec write_message(t(), Message.message()) :: :ok | {:error, any()}
  def write_message(%__MODULE__{device: device}, message) do
    map = Message.message_to_map(message)

    case JsonCodec.encode(map) do
      {:error, reason} ->
        {:error, reason}

      {:ok, json} ->
        write_raw(device, json)
    end
  end

  # ---------------------------------------------------------------------------
  # write_raw/2 — write a raw JSON binary with framing
  # ---------------------------------------------------------------------------

  @doc """
  Write a raw JSON binary string with Content-Length framing. Useful when
  the caller has already encoded the message or wants to write a custom payload.

  Returns `:ok` on success or `{:error, reason}` on failure.

  ## Example

      :ok = Writer.write_raw(writer, ~s({"jsonrpc":"2.0","id":1,"result":42}))
  """
  @spec write_raw(any(), binary()) :: :ok | {:error, any()}
  def write_raw(device, json) when is_binary(json) do
    # The Content-Length value is the BYTE length, not the character count.
    # For ASCII-only JSON this is the same, but for JSON with Unicode (e.g.
    # file paths with non-ASCII characters), byte_size and String.length differ.
    n = byte_size(json)
    header = "Content-Length: #{n}\r\n\r\n"
    payload = header <> json

    case IO.binwrite(device, payload) do
      :ok -> :ok
      {:error, _} = err -> err
    end
  end
end
